crates/build/re_types_builder/src/codegen/rust/deserializer.rs

```diff
@@ -349,24 +348,16 @@ pub fn quote_arrow_deserializer(
 
                         quote! {
                             let #data_dst = {
-                                // NOTE: `data_src_arrays` is a runtime collection of all of the
-                                // input's payload's union arms, while `#type_id` is our comptime union
-                                // arm counter… there's no guarantee it's actually there at
-                                // runtime!
-                                if #data_src_arrays.len() <= #type_id {
-                                    // By not returning an error but rather defaulting to an empty
-                                    // vector, we introduce some kind of light forwards compatibility:
-                                    // old clients that don't yet know about the new arms can still
-                                    // send data in.
-                                    return Ok(Vec::new());
-
-                                    // return Err(DeserializationError::missing_union_arm(
-                                    //     #quoted_datatype, #obj_field_fqname, #type_id,
-                                    // )).with_context(#obj_fqname);
-                                }
-
-                                // NOTE: The array indexing is safe: checked above.
-                                let #data_src = &*#data_src_arrays[#type_id];
+                                // `.child()` will panic if the given `type_id` doesn't exist,
+                                // which could happen if the number of union arms has changed
+                                // between serialization and deserialization.
+                                // There is no simple way to check for this using `arrow-rs`
+                                // (no access to `UnionArray::fields` as of arrow 54:
+                                // https://docs.rs/arrow/latest/arrow/array/struct.UnionArray.html)
+                                //
+                                // Still, we're planning on removing arrow unions entirely, so this is… fine.
+                                // TODO(#6388): stop using arrow unions, and remove this peril
+                                let #data_src = #data_src.child(#type_id).as_ref();
```

Comment:
Why not keep the existing if-statement that helped with forward compat though?

Reply:
There was no obvious way how to make it work. There is no `data_src_arrays` anymore because there is no `.fields()` on `UnionArray`